README.md
# Redmine2 Cookbook

[![Build Status](https://secure.travis-ci.org/aminin/redmine2-cookbook.png?branch=master)](http://travis-ci.org/aminin/redmine2-cookbook)

Installs Redmine v2, a Ruby on Rails ticket tracking and wiki tool

## Requirements

### Chef

* Chef 12.1+

### Platform

Tested on ubuntu 12.04, 14.04

### cookbooks

* postgresql
* [rbenv](https://github.com/RiotGamesCookbooks/rbenv-cookbook)
* [ruby_build](https://github.com/fnichol/chef-ruby_build)
* chef_nginx
* database

## Attributes

| Key                                       | Type    | Description                      | Default                               |
|-------------------------------------------|---------|----------------------------------|---------------------------------------|
| <tt>['redmine']['home']</tt>              | String  | Location for Redmine application | <tt>/home/redmine</tt>                |
| <tt>['redmine']['host']</tt>              | String  | Redmine Domain                   | <tt>redmine.example.com</tt>          |
| <tt>['redmine']['user']</tt>              | String  | Owner of redmine files           | <tt>redmine</tt>                      |
| <tt>['redmine']['prefix']</tt>            | String  | URL path prefix for redmine      | <tt>nil</tt>                          |
| <tt>['redmine']['ruby_version']</tt>      | String  | Redmine Ruby Version             | <tt>1.9.3-p484</tt>                   |
| <tt>['redmine']['version']</tt>           | String  | Redmine version                  | <tt>2.6.0</tt>                        |
| <tt>['redmine']['bundle_exclude']</tt>    | String  | Exclude groups from bundle       | <tt>%w(development test rmagick)</tt> |
| <tt>['redmine']['listen_port']</tt>       | Fixnum  | Port to listen on                | <tt>80</tt>                           |
| <tt>['redmine']['create_db']</tt>         | Bool    | Whether to create database       | <tt>true</tt>                         |
| <tt>['redmine']['ssl_data_bag_name']</tt> | String  | Data bag holding certificate     | <tt>nil</tt>                          |
| <tt>['redmine']['ssl_cert_dir']</tt>      | String  | Directory to put certificate in  | <tt>#{node['nginx']['dir']}/ssl</tt>  |
| <tt>['redmine']['ssl_listen_port']</tt>   | Fixnum  | Port to listen on for HTTPS      | <tt>443</tt>                          |
| <tt>['redmine']['environment']</tt>       | String  | Rails environment to run         | <tt>node.chef_environment</tt>        |
| <tt>['redmine']['db']['type']</tt>        | String  | Type of redmine database         | <tt>postgresql</tt>                   |
| <tt>['redmine']['db']['dbname']</tt>      | String  | Redmine DB name                  | <tt>redmine</tt>                      |
| <tt>['redmine']['db']['username']</tt>    | String  | Redmine DB user                  | <tt>redmine</tt>                      |
| <tt>['redmine']['db']['hostname']</tt>    | String  | Redmine DB host                  | <tt>localhost</tt>                    |
| <tt>['redmine']['db']['password']</tt>    | String  | Redmine DB password              | <tt>123456</tt>                       |
| <tt>['redmine']['init_style']</tt>        | String  | Init system to use for redmine service <tt>upstart</tt>

See mysql recipe for `['mysql']` attributes

## Usage

To install via librarian-chef add to your Cheffile the following lines

```
cookbook 'rbenv', git: 'https://github.com/RiotGamesCookbooks/rbenv-cookbook'
cookbook 'redmine2', git: 'https://github.com/Fatsoma/redmine2-cookbook'
```

and run `librarian-chef install`

Configure your role/node e.g.:

```ruby
{
    redmine: {
        host: 'redmine.dev',
        db: {
            password: '<top-secret1>'
        }
    },
    postgresql: {
        password: {
            postgres: '<top-secret2>' # Need admin access to create redmine DB
        }
    },
    run_list: %w(recipe[postgresql::server] recipe[redmine2])
}
```

## Running tests

```
bundle exec rake foodcritic
bundle exec rake kitchen:all
```

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

## License and Authors

Authors (on Fatsoma):

* Bill Ruddock

Authors (before forking to Fatsoma):

* Anton Minin
* Kazuki Matsuda
* Noah kobayashi


CHANGELOG.md
# Redmine2 CHANGELOG

This file is used to list changes made in each version of the redmine2 cookbook.

## 0.5.0

- Add support for Chef 12
- Drop support for Chef 11
- Drop restriction on mysql cookbook version (with breaking changes to mysql attributes)
- Add documentation for all support redmine attributes
- Add upstart and systemd service scripts. Use upstart by default
- Remove runit service script
- Add attribute `['redmine']['environment']` to set application environment
- Use redmine 3.3.2 and ruby 2.3.1 by default
- Change depend from nginx to chef_nginx cookbook

## 0.4.2

Changes up to 0.4.2 not recorded

## 0.1.0

- [Anton Minin] - Initial release of redmine2

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


test/integration/postgresql/serverspec/redmine_spec.rb
require 'spec_helper'

describe user('redmine') do
  it { should exist }
end

describe user('redmine') do
  it { should have_home_directory '/home/redmine' }
end

describe command('ruby -v') do
  let(:path) { '/opt/rbenv/shims:$PATH' }
  its(:stdout) { should match(/ruby 1\.9\.3p484.*/) }
end

describe package('bundler') do
  let(:path) { '/opt/rbenv/shims:$PATH' }
  it { should be_installed.by('gem') }
end

describe file('/home/redmine/redmine-2.6.0') do
  it { should be_directory }
end

describe file('/home/redmine/redmine') do
  it { should be_linked_to '/home/redmine/redmine-2.6.0' }
end

describe process('ruby') do
  its(:args) { should match(/script\/rails s thin -p 3000/) }
end


test/integration/postgresql/serverspec/postgresql_spec.rb
require 'spec_helper'

describe process('postgres') do
  it { should be_running }
end

describe command('PGPASSWORD=redmine psql -h localhost -U redmine -l >/dev/null') do
  its(:exit_status) { should eq 0 }
end


recipes/mysql.rb
#
# Cookbook Name:: redmine2
# Recipe:: mysql
#
# Copyright 2014, Anton Minin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

mysql_service node[:mysql][:service_name] do
  charset node[:mysql][:charset]
  data_dir node[:mysql][:data_dir]
  error_log node[:mysql][:error_log]
  initial_root_password node[:mysql][:server_root_password]
  mysqld_options node[:mysql][:mysqld_options]
  port node[:mysql][:port]
  run_group node[:mysql][:server_group]
  run_user node[:mysql][:server_user]
  package_version node[:mysql][:server_package_version]
  package_action node[:mysql][:server_package_action]
  action [:create, :start]
  only_if { %w(localhost 127.0.0.1).include? node[:redmine][:db][:hostname] }
end

mysql_client 'default' do
  action :create
end

if [true, 'true'].include? node[:redmine][:create_db]
  include_recipe 'database::mysql'

  connection_info = {
    host:     node[:redmine][:db][:hostname],
    username: 'root',
    password: node[:mysql][:server_root_password]
  }

  mysql_database_user node[:redmine][:db][:username] do
    connection connection_info
    password   node[:redmine][:db][:password]
    action     :create
  end

  mysql_database node[:redmine][:db][:dbname] do
    connection connection_info
    owner node[:redmine][:db][:username]
    encoding 'utf8'
    action :create
  end

  mysql_database_user node[:redmine][:db][:username] do
    connection    connection_info
    database_name node[:redmine][:db][:dbname]
    host          '%'
    privileges    [:all]
    action        :grant
  end
end


recipes/default.rb
#
# Cookbook Name:: redmine2
# Recipe:: default
#
# Copyright 2014, Anton Minin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Create a redmine OS account with the users cookbook
user node[:redmine][:user] do
  comment 'Redmine user'
  shell '/bin/bash'
  home node[:redmine][:home]
  action :create
end

# Create home directory
directory node[:redmine][:home] do
  owner node[:redmine][:user]
  group node[:redmine][:user]
  mode '0755'
end

if node[:redmine][:ruby_version] == 'system'
  bundle_command = 'bundle'
  rake_command = 'rake'
  ruby_command = 'ruby'
else
  # Install ruby with rbenv
  include_recipe 'rbenv::default'
  include_recipe 'ruby_build'

  rbenv_ruby node[:redmine][:ruby_version] do
    global true
  end

  rbenv_gem 'bundler' do
    ruby_version node[:redmine][:ruby_version]
  end

  bundle_command = "#{node[:rbenv][:root]}/shims/bundle"
  rake_command = "#{node[:rbenv][:root]}/shims/bundle exec rake"
  ruby_command = "#{node[:rbenv][:root]}/shims/ruby"
end

# Download archive with source code
bash 'install_redmine' do
  cwd node[:redmine][:home]
  user node[:redmine][:user]
  code <<-EOH
    wget http://www.redmine.org/releases/redmine-#{node[:redmine][:version]}.tar.gz;
    tar -xzf redmine-#{node[:redmine][:version]}.tar.gz
  EOH
  not_if { ::File.exist?("#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/Rakefile") }
end

link "#{node[:redmine][:home]}/redmine" do
  to "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}"
end

# Create database
case node[:redmine][:db][:type]
when 'sqlite'
  include_recipe 'redmine2::sqlite'
when 'mysql'
  include_recipe 'redmine2::mysql'
when 'postgresql'
  include_recipe 'redmine2::postgresql'
end

template "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/config/database.yml" do
  source 'database.yml.erb'
  owner node[:redmine][:user]
  variables database_server: node[:redmine][:db][:hostname]
  mode '0664'
end

# Configure email
template "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/config/configuration.yml" do
  source 'configuration.yml.erb'
  owner node[:redmine][:user]
  variables {}
  mode '0664'
end

# Configure environment.rb
template "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/config/environment.rb" do
  source 'environment.rb.erb'
  owner node[:redmine][:user]
  variables {}
  mode '0644'
end

# Configure custom gems e.g. thin
template "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/Gemfile.local" do
  source 'Gemfile.local.erb'
  owner node[:redmine][:user]
  variables {}
  mode '0664'
end

exclude_db_groups =
  case node[:redmine][:db][:type]
  when 'sqlite'
    %w(mysql postgresql)
  when 'mysql'
    %w(postgresql sqlite)
  when 'postgresql'
    %w(mysql sqlite)
  end
bundle_exclude_groups = node[:redmine][:bundle_exclude] + exclude_db_groups
bundle_install_command = "#{bundle_command} install --without #{bundle_exclude_groups.join(' ')} --path vendor/bundle"

execute bundle_install_command do
  user node[:redmine][:user]
  cwd "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}"
  not_if { ::File.exist?("#{node[:redmine][:home]}redmine-#{node[:redmine][:version]}/db/schema.rb") }
end

execute "RAILS_ENV='production' #{rake_command} db:migrate" do
  user node[:redmine][:user]
  cwd "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}"
  not_if { ::File.exist?("#{node[:redmine][:home]}redmine-#{node[:redmine][:version]}/db/schema.rb") }
end

execute "RAILS_ENV='production' #{rake_command} generate_secret_token" do
  user node[:redmine][:user]
  cwd "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}"
  not_if { ::File.exist?("#{node[:redmine][:home]}redmine-#{node[:redmine][:version]}/config/initializers/secret_token.rb") }
end

case node[:redmine][:init_style]
when 'upstart'
  template '/etc/init/redmine.conf' do
    source 'redmine-upstart.conf.erb'
    variables(
      env_vars: {
        'RAILS_ENV' => node[:redmine][:environment]
      },
      user: node[:redmine][:user],
      app_path: "#{node[:redmine][:home]}/redmine",
      ruby_command: ruby_command
    )
  end

  service 'redmine' do
    provider Chef::Provider::Service::Upstart
    supports status: true, restart: true, reload: true
    action [:start, :enable]
  end
when 'systemd'
  template '/etc/systemd/system/redmine.service' do
    source 'redmine.service.erb'
    variables(
      env_vars: {
        'RAILS_ENV' => node[:redmine][:environment]
      },
      user: node[:redmine][:user],
      app_path: "#{node[:redmine][:home]}/redmine",
      ruby_command: ruby_command
    )
  end

  service 'redmine' do
    provider Chef::Provider::Service::Systemd
    supports status: true, restart: true, reload: true
    action [:start, :enable]
  end
end

certificate_manage node[:redmine][:ssl_data_bag_name].to_s do
  cert_path node[:redmine][:ssl_cert_dir]
  owner node[:nginx][:user]
  group node[:nginx][:user]
  nginx_cert true
  create_subfolders true
  not_if { node[:redmine][:ssl_data_bag_name].nil? }
end

include_recipe 'chef_nginx'

template "#{node[:nginx][:dir]}/sites-available/redmine" do
  source 'nginx-redmine.erb'
  mode 0777
  owner node[:nginx][:user]
  group node[:nginx][:user]
  variables(
    app_path:        "#{node[:redmine][:home]}/redmine",
    server_name:     node[:redmine][:host],
    listen_port:     node[:redmine][:listen_port],
    ssl_listen_port: node[:redmine][:ssl_listen_port],
    ssl_cert:        "#{node[:redmine][:ssl_cert_dir]}/certs/#{node[:fqdn]}.pem",
    ssl_key:         "#{node[:redmine][:ssl_cert_dir]}/private/#{node[:fqdn]}.key"
  )
end

nginx_site 'redmine' do
  enable true
end

# Configure SCM e.g. Git


recipes/plugins.rb
#
# Cookbook Name:: redmine2
# Recipe:: plugins
#
# Copyright 2014, Anton Minin
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

package 'git-core'
package 'unzip'

bundle_command = "#{node[:rbenv][:root]}/shims/bundle"
rake_command = "#{node[:rbenv][:root]}/shims/bundle exec rake"
exclude_db_groups =
  case node[:redmine][:db][:type]
  when 'sqlite'
    %w(mysql postgresql)
  when 'mysql'
    %w(postgresql sqlite)
  when 'postgresql'
    %w(mysql sqlite)
  end
bundle_exclude_groups = node[:redmine][:bundle_exclude] + exclude_db_groups
bundle_install_command = "#{bundle_command} install --without #{bundle_exclude_groups.join(' ')} --path vendor/bundle"

execute bundle_install_command do
  user node[:redmine][:user]
  cwd "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}"
  action :nothing
end

execute "RAILS_ENV='production' #{rake_command} redmine:plugins:migrate" do
  user node[:redmine][:user]
  cwd "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}"
  action :nothing
end

service 'redmine' do
  supports status: true, start: true, stop: true, restart: true
  action :nothing
end

plugins = node[:redmine][:plugins]

if !plugins.nil? && !plugins.empty?

  plugins.each do |plugin|

    case plugin[:type]
    when 'git' then
      rev = plugin[:revision] || 'master'
      git "#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/plugins/#{plugin[:name]}" do
        repository plugin[:source]
        revision rev
        action :sync
        notifies :run, "execute[#{bundle_install_command}]", :delayed
        notifies :run, "execute[RAILS_ENV='production' #{rake_command} redmine:plugins:migrate]", :delayed
        notifies :restart, 'service[redmine]', :delayed
      end

    when 'zip' then
      zipfile = File.basename(plugin[:source])
      # FC041
      bash "Deploy #{plugin[:name]}" do
        cwd '/tmp'
        code <<-EOF
          wget #{plugin[:source]}
          unzip #{zipfile}
          mv #{plugin[:name]} #{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/plugins/#{plugin[:name]}
          rm -rf #{zipfile}
        EOF
        notifies :run, "execute[#{bundle_install_command}]", :delayed
        notifies :run, "execute[RAILS_ENV='production' #{rake_command} redmine:plugins:migrate]", :delayed
        notifies :restart, 'service[redmine]', :delayed
        not_if { ::File.exist?("#{node[:redmine][:home]}/redmine-#{node[:redmine][:version]}/plugins/#{plugin[:name]}") }
      end
    end
  end
end


metadata.rb
name             'redmine2'
maintainer       'Bill Ruddock'
maintainer_email 'bill.ruddock@fatsoma.com'
license          'Apache 2.0'
description      'Installs/Configures redmine 2 or 3'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '0.5.0'

issues_url 'https://github.com/Fatsoma/redmine2-cookbook/issues'
source_url 'https://github.com/Fatsoma/redmine2-cookbook'

chef_version '~> 12'

conflicts 'redmine'

depends 'rbenv'      # https://github.com/RiotGamesCookbooks/rbenv-cookbook
depends 'ruby_build' # https://github.com/fnichol/chef-ruby_build
depends 'chef_nginx'
depends 'database'
depends 'postgresql'
depends 'mysql'
depends 'sqlite'
depends 'certificate'
depends 'iptables'

supports 'ubuntu'

attribute 'redmine',
          display_name:   'Redmine Hash',
          description:    'Hash of Redmine attributes',
          type:           'hash'

attribute 'redmine/home',
          display_name:    'Redmine Directory',
          description:     'Location for Redmine application',
          default:         '/home/redmine',
          recipes:         ['redmine2::default']

attribute 'redmine/host',
          display_name:    'Redmine Domain',
          description:     'Redmine Domain',
          default:         'redmine.example.com',
          recipes:         ['redmine2::default']

attribute 'redmine/user',
          display_name:    'Redmine User',
          description:     'Owner of redmine files',
          default:         'redmine',
          recipes:         ['redmine2::default']

attribute 'redmine/prefix',
          display_name:    'Redmine URL prefix',
          description:     'URL path prefix for Redmine application',
          default:         nil,
          recipes:         ['redmine2::default']

attribute 'redmine/ruby_version',
          display_name:    'Redmine Ruby Version',
          description:     'Version of Ruby to run Redmine',
          default:         '1.9.3-p484',
          recipes:         ['redmine2::default']

attribute 'redmine/version',
          display_name:    'Redmine version',
          description:     'Redmine version',
          default:         '2.6.0',
          recipes:         ['redmine2::default']

attribute 'redmine/bundle_exclude',
          display_name:    'Bundle exclude groups',
          description:     'Exclude groups from bundle',
          default:         %w(development test rmagick),
          recipes:         ['redmine2::default']

attribute 'redmine/listen_port',
          display_name:    'Listen port',
          description:     'Port to listen on in nginx',
          default:         80,
          recipes:         ['redmine2::default']

attribute 'redmine/create_db',
          display_name:    'Create DB on install',
          description:     'Whether to create DB',
          default:         'true',
          recipes:         ['redmine2::default']

attribute 'redmine/ssl_data_bag_name',
          display_name:    'SSL/TLS certificate data bag name',
          description:     'Data bag holding SSL/TLS certificate',
          default:         nil,
          recipes:         ['redmine2::default']

attribute 'redmine/ssl_cert_dir',
          display_name:    'SSL/TLS certificate directory',
          description:     'Directory to install SSL/TLS certificate to',
          default:         '/etc/nginx/ssl',
          recipes:         ['redmine2::default']

attribute 'redmine/ssl_listen_port',
          display_name:    'HTTPS Listen port',
          description:     'Port to listen on in nginx for HTTPS',
          default:         443,
          recipes:         ['redmine2::default']

attribute 'redmine/environment',
          display_name:    'Rails environment',
          description:     'Environment to run rails in',
          default:         'node.chef_environment',
          recipes:         ['redmine2::default']

attribute 'redmine/db',
          display_name:    'Redmine DB Hash',
          description:     'Hash of redmine database attributes',
          type:            'hash'

attribute 'redmine/db/type',
          display_name:    'Redmine DB type',
          description:     'Type of redmine database',
          choice:          %w(sqlite postgresql mysql),
          default:         'postgresql',
          recipes:         ['redmine2::default']

attribute 'redmine/db/dbname',
          display_name:    'Redmine DB name',
          description:     'Redmine DB name',
          default:         'redmine',
          recipes:         ['redmine2::default']

attribute 'redmine/db/username',
          display_name:    'Redmine DB user',
          description:     'Redmine DB user',
          default:         'redmine',
          recipes:         ['redmine2::default']

attribute 'redmine/db/hostname',
          display_name:    'Redmine DB host',
          description:     'Redmine DB host',
          default:         'localhost',
          recipes:         ['redmine2::default']

attribute 'redmine/db/password',
          display_name:    'Redmine DB password',
          description:     'Redmine DB password',
          default:         '123456',
          recipes:         ['redmine2::default']

attribute 'redmine/init_style',
          display_name:    'Redmine service init style',
          description:     'Init system to use',
          choice:          %w(upstart systemd),
          default:         'upstart',
          recipes:         ['redmine2::default']
